### DISCUSSION TITLE
Help wanted: JSON:API

### DISCUSSION THREAD

**Discussion by okuramasafumi** [okuramasafumi/alba#217] [General]
On [jsonapi branch](https://github.com/okuramasafumi/alba/tree/jsonapi) I'm developing a [JSON:API](https://jsonapi.org/) compatibility but there's an issue.
It's hard to confirm Alba is actually compatible with JSON:API. There's no test suite so I copied some specs from https://github.com/jsonapi-serializer/jsonapi-serializer but we need more specs.

Could anyone add some specs for JSON:API?

**Comment by jrochkind** [comment]
Is it helpful to realize there's a JSON:API [JSON Schema](https://jsonapi.org/schema), and you could use the ruby [json_schema](https://github.com/voxpupuli/json-schema) gem to validate any JSON against the JSON:API JSON schema?  (Although to be fair, [json:api FAQ](https://jsonapi.org/faq/#is-there-a-json-schema-describing-json-api) says "Validation will not yield false negatives, but could yield false positives for the sake of flexibility.")

That might be one way to verify that any given output is legal JSON:API, although there may be more to verifying compatibility than that. 

I am definitely interested in json:api support here. Looking at the diff between your branch and master, I haven't quite wrapped my head around what's going on. 

Can you say more about where you are stuck, or if there's any way for anyone else to help?

I would also be interested in the possiiblity of using the same serializer to serialize as JSON:API or "plain", not sure if that's feasible or not.

**Reply by okuramasafumi** [reply]
> I am definitely interested in json:api support here. Looking at the diff between your branch and master, I haven't quite wrapped my head around what's going on.

Thank you, what I want to achieve is to use Alba's existing functionalities as much as possible. JSONAPI module "inherits" from `Alba::Resource` module and expands its functionalities.

> Can you say more about where you are stuck, or if there's any way for anyone else to help?

I got stuck due to the lack of "official" specs that ensure Alba is compatible with JSONAPI. I think it kinds works now, but will have many problems when it goes production.

**Reply by ngouy** [reply]
What do you think of having a list/table of features that JSON:API "requires", then having 2 columns:
- specs available: we have made some specs that fits 100% what JSON:API is asking for
- supported: we made the code to make column 1 happy. So we can say feature x is supported

And at least from there we could know how far we are, or what coverage we have for supporting JSON:API specs

**Comment by hwo411** [comment]
Hi @okuramasafumi . Is it still relevant for you? It's been almost 3 years since this discussion and I see that the branch pretty outdated. But since `jsonapi-serializer` is kinda dead, we have a desire to move away from it in the nearest future and Alba seems to be a good candidate given the JSON API functionality is implemented.

Would you like to resume the work and collaborate on the JSON API integration?

**Reply by okuramasafumi** [reply]
Hi @hwo411 although I don't use JSON:API spec, this feature is still nice-to-have for me.
As I mentioned in https://github.com/okuramasafumi/alba/discussions/217#discussion-4243632 the main issue is the lack of the test suite.
Things have changed and now AI helps us to build effective test suite, so implementing reliable JSON:API support might be easier.

**Reply by hwo411** [reply]
It's a good point regarding AI!

I also think that we can test JSON API implementation with our codebase, since transition shouldn't be difficult and can be done one by one for each serializer. And we have quite a good test coverage to detect issues and differences.

**Reply by okuramasafumi** [reply]
Cool, I'll rebase this old branch and see what we can do!

**Reply by dodeja** [reply]
I might be able to help aswell
